src/libsystemd/sd-daemon/sd-daemon.c

```diff
@@ -551,6 +553,39 @@ _public_ int sd_pid_notify_with_fds(
         return r;
 }
 
+_public_ int sd_notify_barrier(int unset_environment, uint64_t timeout) {
+        _cleanup_close_pair_ int pipe_fd[2] = { -1, -1 };
+        struct timespec ts;
+        int r;
+
+        if (pipe2(pipe_fd, O_CLOEXEC) < 0) {
+                r = -errno;
+                goto finish;
+        }
+
+        struct pollfd pfd = {
+                .fd = pipe_fd[0],
+                /* POLLHUP is implicit */
+                .events = 0,
+        };
+
+        r = sd_pid_notify_with_fds(0, unset_environment, "BARRIER=1", &pipe_fd[1], 1);
+        if (r <= 0)
+                goto finish;
+
+        pipe_fd[1] = safe_close(pipe_fd[1]);
+
+        if (ppoll(&pfd, 1, timeout == UINT64_MAX ? NULL : timespec_store(&ts, timeout), NULL) >= 0) {
+                r = 1;
```

Comment:
actually, we should probably proagate some suitable error when the timeout is hit. e.g:

```
r = ppoll();
if (r < 0)
        return -errno;
if (r == 0)
        return -ETIMEDOUT;
return 0;
```